src/Stack/Setup.hs

```diff
@@ -669,6 +678,84 @@ ensureCompiler sopts = do
       , cpGlobalDB' = globaldb
       }
 
+buildGhcFromSource :: forall env.
+   ( HasTerm env
+   , HasProcessContext env
+   , HasConfig env
+   ) => Memoized SetupInfo -> [Tool] -> CompilerGitBuild
+   -> RIO env (Either (ActualCompiler, Arch, Path Abs Dir) Tool, CompilerBuild)
+buildGhcFromSource getSetupInfo' installed (CompilerGitBuild repo flavour) = do
+   config <- view configL
+   let compilerTool = ToolGhcGit (repoCommit repo) flavour
+
+   -- detect when the correct GHC is already installed
+   if compilerTool `elem` installed
+     then return (Right compilerTool,CompilerBuildStandard)
+     else do
+       -- clone the repository and execute the given commands
+       Pantry.withRepo repo $ do
+         threads <- view $ configL.to configJobs
+         let
+           hadrianArgs = fmap T.unpack
+               [ hadrianCmd
+               , "-c"                    -- run ./boot and ./configure
+               , "-j" <> tshow threads   -- parallel build
+               , "--flavour=" <> flavour -- selected flavour
+               , "binary-dist"
+               ]
+           hadrianCmd
+             | osIsWindows = "./hadrian/build.stack.bat"
+             | otherwise   = "./hadrian/build.stack.sh"
+
+         logSticky $ "Building GHC from source with `"
+            <> RIO.display flavour
+            <> "` flavour. It can take a long time (more than one hour)..."
+
+         -- RIO.Process doesn't wrap process' "shell".
+         -- Instead we use "proc" with the "sh" command
+         void (proc "sh" hadrianArgs readProcess_)
```

Comment:
`readProcess_` here will cause no output to be displayed to the user during the installation, and will take up a large amount of memory. `runProcess_` may be a better choice, perhaps with redirecting `stdout` to `stderr`.

Reply:
I use ``runprocess_`` now. It is more verbose given how Hadrian is but I'm quite used to it so I don't now what the users of this feature would expect.